### PR TITLE
Update rspec.ps1 to copy test case from RSPEC

### PR DESCRIPTION
**Pull request by sebastien-marichal**
As per the new rule specification process, this PR update `rspec.ps1` to copy all test cases from the given rule and branch.

### CODE REVIEW

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Looks good, I have only 1 concern

File: scripts/rspec/rspec.ps1 [hunk 1]
```diff
@@ -32,6 +32,15 @@ param (
 Set-StrictMode -version 1.0
 $ErrorActionPreference = "Stop"
 $RuleTemplateFolder = "${PSScriptRoot}\\rspec-templates"
+# Based on RuleApiCache.computeDefaultCachePath
+# https://github.com/SonarSource/sonar-rule-api/blob/2323b7313c76e7adc2b7df037e96510b90660292/src/main/java/com/sonarsource/ruleapi/utilities/RuleApiCache.java#L20-L25
+$RspecRepositoryPath = "${Env:USERPROFILE}\\.sonar\\rule-api\\rspec"
+if (! [string]::IsNullOrEmpty($Env:SONAR_USER_HOME))
```
[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply scripts/rspec/rspec.ps1:38]: `$RspecRepositoryPath` will be set to "${Env:USERPROFILE}\\.sonar\\rule-api\\rspec".
I reproduce the behavior in `rule-api`.

File: scripts/rspec/CopyTestCasesFromRspec.ps1 [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# 3. The test case file name does not contain a dot, but the test case file is in a subfolder, e.g. "MyTestCase\Scenario.razor"
+#    Then the scenario name is the file name, e.g. "Scenario", ignoring the folder name
+# 4. The test case file name does not contain a dot and the test case file is in the root folder, e.g. "MyTestCase.razor"
+#    Then the scenario name is empty, e.g. ""
+#
+# Example:
+#   1. MyTestCase.Scenario.razor            =>  RuleName.Scenario.razor
+#   2. MyTestCase\MyTestCase.Scenario.razor =>  RuleName.Scenario.razor
+#   3. MyTestCase\Scenario.razor            =>  RuleName.Scenario.razor
+#   4. MyTestCase.razor                     =>  RuleName.razor
```
[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply scripts/rspec/CopyTestCasesFromRspec.ps1:17]: If there is no `.` in the file name or the rule is not within a folder, no scenario will be found.
If you have multiple files with no scenario, for now, they will override each other.

I think we need to think a bit more about this.

File: scripts/rspec/rspec.ps1 [hunk 3]
```diff
@@ -224,6 +233,7 @@ if ($ClassName -And $RuleKey) {
     elseif ($Language -eq "vbnet") {
        GenerateRuleClassesVB
     }
+    CopyTestCasesFromRspec $ClassName "${RspecRepositoryPath}\\rules\\${RuleKey}\\csharp" $TestCasesFolder
```
Comment (sebastien-marichal) [inline scripts/rspec/rspec.ps1:236]: `vbnet` should be handled here as well.

File: scripts/rspec/CopyTestCasesFromRspec.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply scripts/rspec/CopyTestCasesFromRspec.ps1:17]: As discussed, we will add a suffix counter to the filename to ensure that files are not overridden.

File: scripts/rspec/rspec.ps1 (hunk 3 above)
Comment (sebastien-marichal) [inline scripts/rspec/rspec.ps1:236]: `CopyTestCasesFromRspec` should be called before generating the default TestCases file.
Update the TestCase file creation to not be generated if there is already a TestCase with the same name.

File: scripts/rspec/CopyTestCasesFromRspec.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply scripts/rspec/CopyTestCasesFromRspec.ps1:17]: Should we document this behavior in the header?

[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply scripts/rspec/CopyTestCasesFromRspec.ps1:17]: We should, I will add it!

File: scripts/rspec/CopyTestCasesFromRspec.ps1 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        elseif ($_.Directory.FullName -ne $(Convert-Path $RspecRulePath))
+        {
+            $scenario = "." + $_.BaseName
+        }
+
+        $outputPath = "${OutputFolder}\$FileName$scenario$($_.Extension)"
+
+        if (Test-Path -Path $outputPath -PathType Leaf)
+        {
+            $count = (Get-ChildItem -Path $OutputFolder -Name | Where-Object { $_.StartsWith("$FileName$scenario") }).Count
```
Comment (cristian-ambrosini-sonarsource) [inline scripts/rspec/CopyTestCasesFromRspec.ps1:50]: I think we should check the extension too. We need to make sure (maybe a test?) that: 
[...]
`FileName.Scenario.1.cs`
[...]
`FileName.Scenario.3.vb`
The following `.cs` file with the same name will be named `FileName.Scenario.2.cs`

File: scripts/rspec/rspec.ps1 [hunk 5]
```diff
@@ -94,7 +108,11 @@ function GenerateRuleClassesVB() {
     }
 
     $FilesMap["Rule.VB.cs"] = "${RulesFolderVB}\\${ClassName}.cs"
-    $FilesMap["TestCase.VB.vb"] = "${TestCasesFolder}\\${ClassName}.vb"
+
+    if (-Not (Test-Path -Path "${TestCasesFolder}\\${className}.vb" -PathType Leaf))
```
Comment (cristian-ambrosini-sonarsource) [inline scripts/rspec/rspec.ps1:112]: Is it case-insensitive?
```suggestion
    if (-Not (Test-Path -Path "${TestCasesFolder}\\${ClassName}.vb" -PathType Leaf))
```

**Review by cristian-ambrosini-sonarsource — CHANGES_REQUESTED** [review]
Almost there!

**Review by cristian-ambrosini-sonarsource — COMMENTED** [review]
Please remember also to document the new behavior on the `README.md` situated in the same script/rspec folder

File: scripts/rspec/rspec.ps1 (hunk 5 above)
Comment (sebastien-marichal) [reply scripts/rspec/rspec.ps1]: It is not case-sensitive, I updated it anyway so it is the same in the function.

File: scripts/rspec/rspec.ps1 [hunk 6]
```diff
@@ -76,9 +85,14 @@ function GenerateRuleClassesCS() {
     $FilesMap = @{
         "Rule.CS.cs"     = "${RulesFolderCS}\\${className}.cs"
         "Test.CS.cs"     = "${RulesFolderTests}\\${className}Test.cs"
-        "TestCase.CS.cs" = "${TestCasesFolder}\\${className}.cs"
     }
-    WriteClasses $FilesMap $ClassName
+
+    if (-Not (Test-Path -Path "${TestCasesFolder}\\${className}.cs" -PathType Leaf))
+    {
+        $FilesMap["TestCase.CS.cs"] = "${TestCasesFolder}\\${className}.cs"
```
Comment (cristian-ambrosini-sonarsource) [inline scripts/rspec/rspec.ps1:92]: Super nitpick (optional): I would have renamed `className` into `ClassName` on the whole function, for consistency with GenerateRuleClassesVB

**Review by cristian-ambrosini-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**